### PR TITLE
Kingston changes after 3:20 pm

### PR DESCRIPTION
**Pull request by ErikCald**
- [x] Climber was incorrectly set in NewRobotContainer. Fixed that and moved it to leftTrigger.
- [x] Ejecting notes out the intake side with a button was improved
- [x] Working vision align and shot. Though too loud, the DB is measured to be above safety regulation.
- [x] Bling informs drivers when the note is inside the intake
- [x] Improvement to swerve encoder syncing which slightly improves the swerve rotation drift

### CODE REVIEW

File: src/main/java/frc/robot/commands/TeleopSwerve.java [hunk 1]
```diff
@@ -71,6 +77,32 @@ protected double calculateRotationVal() {
     return rotationLimiter.calculate(rotationVal);
   }
 
+  // protected double calculateRotationVal() {
```
Comment (weizhang-2706) [inline src/main/java/frc/robot/commands/TeleopSwerve.java:80]: If this method is not useful for later, suggest to delete it.

**Review by weizhang-2706 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/TeleopSwerve.java [hunk 2]
```diff
@@ -81,6 +113,9 @@ public void execute() {
             calculateRotationVal()),
             isFieldRelative,
         true);
-  }
 
+    // if (rotationVal != 0.0 || Math.toDegrees(s_Swerve.getRobotRelativeSpeeds().omegaRadiansPerSecond) < 10) {
```
Comment (weizhang-2706) [inline src/main/java/frc/robot/commands/TeleopSwerve.java:117]: the same comments as the previous one. Clean up.

File: src/main/java/frc/robot/subsystems/SwerveSubsystem.java [hunk 3]
```diff
@@ -152,13 +153,14 @@ private SwerveSubsystem() {
...
-      ChassisSpeeds speeds, boolean fieldRelative, boolean isOpenLoop) {
+  public void drive(ChassisSpeeds speeds, boolean fieldRelative, boolean isOpenLoop, boolean flipForAlliance) {
+    Rotation2d heading = flipForAlliance ? rotateForAlliance(getHeading()) : getHeading();
+
     SwerveModuleState[] swerveModuleStates =
     Config.Swerve.swerveKinematics.toSwerveModuleStates(
       // ChassisSpeeds.discretize(
         fieldRelative ? 
-            ChassisSpeeds.fromFieldRelativeSpeeds(speeds, rotateForAlliance(getHeading())) :
+            ChassisSpeeds.fromFieldRelativeSpeeds(speeds, heading) :
```
Comment (weizhang-2706) [inline src/main/java/frc/robot/subsystems/SwerveSubsystem.java:163]: why do we need to take account for blue/red alliance here?
Auto starting position will set it correctly.

File: src/main/java/frc/robot/subsystems/SwerveSubsystem.java [hunk 4]
```diff
@@ -348,7 +355,7 @@ public void periodic() {
     // If the robot isn't moving synchronize the encoders every 100ms (Inspired by democrat's SDS
     // lib)
     // To ensure that everytime we initialize it works.
-    if (!isChassisMoving(0.01)) {
+    if (!isChassisMoving(0.01) && !areModulesRotating(2)) {
```
Comment (weizhang-2706) [inline src/main/java/frc/robot/subsystems/SwerveSubsystem.java:358]: This is a reasonable fix.

File: src/main/java/frc/robot/subsystems/SwerveSubsystem.java (hunk 3 above)
Comment (ErikCald) [reply src/main/java/frc/robot/subsystems/SwerveSubsystem.java:163]: On the blue alliance, forwards is defined as a heading of 0. On the red alliance, forwards is defined as a heading of 180 degrees (which is set by auto). But the drivers still want forwards to be away from them. So on the red alliance, the driver controls need to rotate the heading by 180 degrees to account for this.

File: src/main/java/frc/robot/commands/TeleopSwerve.java (hunk 1 above)
Comment (ErikCald) [reply src/main/java/frc/robot/commands/TeleopSwerve.java:80]: This was an attempted implementation by Khalil and Fernando which wasn't working so we had to abandon it for new market. It should ideally be implemented for north bay though so I think leaving it in is fine.

It should use the gyro to keep the heading constant whenever the driver doesn't want the robot to rotate.

Comment (weizhang-2706) [reply src/main/java/frc/robot/commands/TeleopSwerve.java:80]: Is this feature similar to the one we removed, which is discretize?

Comment (ErikCald) [reply src/main/java/frc/robot/commands/TeleopSwerve.java:80]: No discretize tries to solve a similar but different problem.

Discretize tries to solve the issue where when driving forwards and rotating, the robot drives in an arc.

But this is trying to solve the issue where due to many different factors, the rotation can slowly move over time so it uses pid to maintain the current heading.

File: src/main/java/frc/robot/subsystems/SwerveSubsystem.java (hunk 3 above)
Comment (weizhang-2706) [reply src/main/java/frc/robot/subsystems/SwerveSubsystem.java:163]: FieldRelativeSpeeds: forward means away from our alliance wall. 
Agreed, for teleop, we do need to adjust by ourselves. In auto, pathPlanner takes care.
"On the blue alliance, forwards is defined as a heading of 0." This is set by PathPlanner?

Comment (ErikCald) [reply src/main/java/frc/robot/subsystems/SwerveSubsystem.java:163]: We draw paths for the blue alliance, pathplanner mirrors them to the red alliance. PathPlanner resets our odometry and has defined "The origin of the field coordinate system will remain on the blue side." This means that, even when on the red alliance, driving towards a heading of 0 will always be driving away from the blue alliance station wall.

But the drivers want the robot to drive the robot according to their perspective, which on the red side is 180 degrees opposite of what odometry is set to by pathplanner. So to fix this, a bunch of headings have to be rotated by 180 degrees when on the red side.

Comment (weizhang-2706) [reply src/main/java/frc/robot/subsystems/SwerveSubsystem.java:163]: yeah, I found the words in pathPlanner, i.e. "The origin of the field coordinate system will remain on the blue side."
Thanks for explanation.

**Review by weizhang-2706 — APPROVED** [review]

**Review by emilycxliu — APPROVED** [review]

**Pull request merged**